lms/djangoapps/course_home_api/outline/views.py

```diff
@@ -520,15 +513,55 @@ def mark_complete_recursive(self, block):
         """
         Mark blocks as complete or not complete based on the completions_dict.
         """
+        if not block:
+            return block
+
         if 'children' in block:
-            block['children'] = [self.mark_complete_recursive(child) for child in block['children']]
-            block['complete'] = all(
-                child['complete'] for child in block['children'] if child['type'] in self.completable_block_types
-            )
+            block['children'] = [self.mark_complete_recursive(child) for child in block['children'] if child]
+            completable_children = self.get_completable_children(block)
+            block['complete'] = all(child['complete'] for child in completable_children)
+            block['completion_stat'] = self.get_block_completion_stat(block, completable_children)
         else:
-            block['complete'] = self.completions_dict.get(block['id'], False)
+            # If the block is a course, chapter, sequential, or vertical, without children,
+            # it should be completed by default.
+            completion = self.completions_dict.get(block['id'], 0)
+            block['complete'] = bool(completion) or block['type'] in ['course', 'chapter', 'sequential', 'vertical']
+            block['completion_stat'] = self.get_block_completion_stat(block, completable_children=[])
+
         return block
 
+    def get_block_completion_stat(self, block, completable_children):
+        """
+        Get the completion status of a block.
+        """
+        block_type = block['type']
+        completable_children_num = len(completable_children)
+
+        if block_type in ['course', 'sequential']:
+            completion = sum(child['complete'] for child in completable_children)
+        elif block_type == 'chapter':
+            # For sections, we have to count the status on the number of completed units
+            # and, accordingly, the number of units in it.
+            completion = sum(child['completion_stat']['completion'] for child in completable_children)
+            completable_children_num = sum(
+                child['completion_stat']['completable_children'] for child in completable_children
+            )
+        elif block_type == 'vertical':
+            completion = sum(child['completion_stat']['completion'] for child in completable_children)
```

Comment:
I think this whole piece needs better comments as to why we're treating these types differently. What is `complete` vs. `completion_stat:completion`? Why are `course` and `sequential` treated the same way, but `chapter` is so different?

Reply:
I've updated the docsting for this method.

`completion`(float from 0 to 1 meaning the percentage of completion) and `complete`(bool value of completion) are the names that are used by default in the `BlockCompletionTransformer`.

The `chapter` has a different calculating, due to a request from the product team, you can read more here:
https://openedx.slack.com/archives/C06A9ES6QUU/p1714586420150249?thread_ts=1714058363.124469&cid=C06A9ES6QUU